test/test-remote/testutils/deployment.ts

```diff
@@ -0,0 +1,283 @@
+        `Try to delete ${r.constructor.name}: ${r.namespace}/${r.name}`
+      );
+      await r.delete();
+    } catch (e) {
+      const err = kubernetesError(e);
+      if (err.statusCode === 404) {
+        log.info("already doesn't exist");
+      } else {
+        throw e;
+      }
+    }
+  }
+}
+
+// Waits for all Deployments/DaemonSets/StatefulSets in the provided list to be Running/Ready.
+// This may be called after deployAll.
+// Note that the specified K8sResources will not be updated.
+export async function waitForAllReady(
+  kubeConfig: KubeConfig,
+  resources: K8sResource[],
+  // If there is a flake with pulling a container image, the pull timeout is 5 minutes.
+  // Give a couple minutes margin to allow for flakes and for the pods to become ready.
+  maxWaitSeconds = 7 * 60
+) {
+  // The provided resources may contain ConfigMaps/etc.
+  // Don't worry about those, just focus on the three types that involve deploying pods.
+  const daemonSets = resources.filter(r => isDaemonSet(r)).map(r => r as DaemonSet);
+  const deployments = resources.filter(r => isDeployment(r)).map(r => r as Deployment);
+  const statefulSets = resources.filter(r => isStatefulSet(r)).map(r => r as StatefulSet);
+
+  const sm = createSagaMiddleware({ onError: function (e: Error, detail: any) {
+    log.error("error seen by saga middleware:\n%s", e.stack);
+    if (detail && detail.sagaStack !== undefined) {
+      log.error("saga stack: %s", detail.sagaStack);
+    }
+    throw Error("exiting after saga error");
+  }});
+  createStore(rootReducer, applyMiddleware(sm));
+  await sm.run(function* () {
+    yield waitForAllReadyImpl(kubeConfig, daemonSets, deployments, statefulSets, maxWaitSeconds);
+  }).toPromise();
+}
+
+function* waitForAllReadyImpl(
+  kubeConfig: KubeConfig,
+  daemonSets: DaemonSet[],
+  deployments: Deployment[],
+  statefulSets: StatefulSet[],
+  maxWaitSeconds: number
+) {
+  //@ts-ignore: TS7075 generator lacks return type (TS 4.3)
+  const informers = yield fork(runInformers, kubeConfig);
+
+  yield call(blockUntilCacheHydrated);
+
+  const { timeout } = yield race({
+    upgrade: call(
+      waitForReady,
+      {
+        originalDaemonSets: daemonSets,
+        originalDeployments: deployments,
+        originalStatefulSets: statefulSets
+      }
+    ),
+    timeout: delay(maxWaitSeconds * 1000)
+  });
+
+  yield cancel(informers);
+  if (timeout) {
+    throw Error(`Timed out waiting for resources to be ready after ${maxWaitSeconds}s`);
+  }
+}
+
+const rootReducers = {
+  kubernetes: combineReducers({
+    cluster: combineReducers({
+      DaemonSets: daemonSetsReducer,
+      Deployments: deploymentsReducer,
+      StatefulSets: statefulSetsReducer
+    })
+  })
+};
+const rootReducer = combineReducers(rootReducers);
+type State = ReturnType<typeof rootReducer>;
+
+function* runInformers(
+  kubeConfig: KubeConfig
+): Generator<
+  ChannelTakeEffect<void | unknown> | PutEffect | CancelledEffect,
+  void,
+  Action
+> {
```

Comment:
When I see these return types (and I had to write them, too, at some point here and there) and some other details I am always a little puzzled by the complexity of having fundamentally diverging programming paradigms in the same code base. Curious how all of that would look without redux-saga. Redux-saga is nice itself but it does not integrate perfectly with the rest of the tooling.

I tried to remove/reduce usage of redux-saga in installer and uninstaller in those places I touched.

https://github.com/opstrace/opstrace/issues/26
https://github.com/opstrace/opstrace/issues/198
https://github.com/opstrace/opstrace/issues/236
https://github.com/opstrace/opstrace-prelaunch/issues/1445